azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import abc
+import six
+from collections import deque
+from . import pipeline_ops
+from . import pipeline_events
+from . import mqtt_topic_helper
+from azure.iot.device.iothub.models import Message
+from azure.iot.device.common.transport.mqtt.mqtt_provider import MQTTProvider
+from azure.iot.device.common.transport import constant
+
+logger = logging.getLogger(__name__)
+
+# TODO: standardize on {} instead of %s
+# TODO: standardize logs with {}({}) prefix
+
+
+@six.add_metaclass(abc.ABCMeta)
+class PipelineStage(object):
+    """
+    Base class representing a stage in the processing pipeline.  Each stage is responsible for receiving
+    PipelineOperation objects from the top, possibly processing them, and possibly passing them down.  It
+    is also responsible for receiving PipelineEvent objects from the bottom, possibly processing them, and
+    possibly passing them up.  In both cases (PipelineOperation and PipelineEvent), the default behavior
+    should be to pass along an operation or event that you don't handle.
+    """
+
+    def __init__(self):
+        self.name = self.__class__.__name__
+        self.next = None
+        self.previous = None
+        self.pipeline_root = None
+
+    def run_op(self, op):
+        """
+        Run the given operation.  This is the public function that outside callers would call to run an
+        operation.  Derived classes should override the _run_op function to impliment stage-specific behavior.
+        """
+        try:
+            self._run_op(op)
+        except Exception as e:
+            logger.error(msg="Error in _run_op call", exc_info=e)
+            op.error = e
+            self.complete_op(op)
+
+    def run_ops_serial(self, *args, **kwargs):
```

Comment:
>run_ops_serial [](start = 8, length = 14)

I'm sure I'm missing something, but I'm not clear on why run_ops_serial is necessary, when we already have ops that chain, what with the .next attribute and the continue_op() method and all.

Reply:
Different concepts.  The .next attribute is the next stage in the pipeline.  continue_op() passes the operation down to the next stage.  Both of these affect how a single op goes through the pipeline.

run_ops_serial is about running multiple ops through the pipeline.  It takes an array of ops.  It passes the first one down, waits for it to travel through the pipeline and eventually complete.  After the first op completes, run_ops_serial passes the second op down the pipeline, and waits for it to complete, and so on.  We could do this with callbacks on each operation, but I found myself repeating that same pattern over and over so I made a helper.

---
In reply to: [279659606](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279659606) [](ancestors = 279659606)